src/NHibernate/Util/AsyncSemaphore.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace NHibernate.Util
+{
+	public class AsyncSemaphore
+	{
+		private static readonly Task completed = Task.FromResult(true);
+		private readonly Queue<TaskCompletionSource<bool>> waiters = new Queue<TaskCompletionSource<bool>>();
+		private int currentCount;
+
+		public AsyncSemaphore(int initialCount)
+		{
+			if (initialCount < 0) throw new ArgumentOutOfRangeException("initialCount");
+			currentCount = initialCount;
+		}
+		public Task WaitAsync()
```

Comment:
The lack of cancellation token here is intentional I think, and should not be "fixed".

Better have an empty line between methods by the way.